File: java/com/google/copybara/util/console/FileConsole.java
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.util.console;

import com.google.common.base.Preconditions;
import com.google.common.flogger.FluentLogger;
import com.google.common.util.concurrent.ListeningScheduledExecutorService;
import com.google.common.util.concurrent.MoreExecutors;
import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.google.copybara.util.console.Message.MessageType;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.Writer;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Duration;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;
import javax.annotation.Nullable;

/**
 * A {@link Console} capable of writing the output into a file and to a delegate console.
 *
 * <p>If any of the file operations fail, the console won't try to write to the file anymore and
 * will only call to the delegate console.
 *
 * <p>Caller is responsible for closing this console to free resources.
 *
 * <p> The console can be configured to flush on a fixed rate intervals.
 */
public class FileConsole extends DelegateConsole {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
  private static final DateTimeFormatter DATE_PREFIX_FMT =
      DateTimeFormatter.ofPattern("MMdd HH:mm:ss.SSS");

  protected final Path filePath;
  private final ListeningScheduledExecutorService flushingExecutor;
  private final ExecutorService loggingExecutor;

  private boolean failed = false;
  @Nullable private Writer writer;


  /**
   * Creates a new {@link FileConsole}.
   *
   * @param delegate A delegate console
   * @param filePath A file path to write to. The parent directories must be created in advance.
   * @param consoleFlushRate How often (in number of messages) to flush this file console.
   */
  public FileConsole(Console delegate, Path filePath, Duration consoleFlushRate) {
    super(delegate);
    this.filePath = Preconditions.checkNotNull(filePath);
    ThreadFactory threadFactory = new ThreadFactoryBuilder()
        .setNameFormat("File logging thread %d")
        .setDaemon(true)
        .setUncaughtExceptionHandler(
            (t, e) -> {
              logger.atSevere().withCause(e).log(
                  "Thread %s threw an unhandled exception: %s", t, e.getMessage());
              System.exit(31 /*ExitCode.INTERNAL_ERROR*/);
            })
        .build();
    loggingExecutor = Executors.newSingleThreadExecutor(threadFactory);
    if (consoleFlushRate.isNegative() || consoleFlushRate.isZero()) {
      flushingExecutor = null;
    } else {
      flushingExecutor = MoreExecutors.listeningDecorator(
          Executors.newSingleThreadScheduledExecutor(threadFactory));
      //noinspection unused : Ignored as this is best effort
      Future<?> ignored = flushingExecutor.scheduleAtFixedRate(() -> {
        logger.atInfo().log("Executing console flush");
        flush();
      }, consoleFlushRate.toMillis(), consoleFlushRate.toMillis(), TimeUnit.MILLISECONDS);
    }
  }

  @Override
  protected void handleMessage(MessageType type, String message) {
    Writer writer = getWriter();
    if (writer == null) {
      return;
    }
    // Submitting to a shut-down executor throws, so this has to be synchronized to avoid shutting
    // down in-between from another thread.
    synchronized (loggingExecutor) {
      if (loggingExecutor.isShutdown()) {
         return;
      }
      //noinspection unused : Ignored as this is best effort
      Future<?> ignored = loggingExecutor.submit(() -> doWrite(writer,
            String.format("%s %s: %s\n",
                ZonedDateTime.now(ZoneId.systemDefault()).format(DATE_PREFIX_FMT), type, message)));
    }
  }

  private void doWrite(Writer writer, String s) {
    if (failed) {
      return;
    }
    try {
      writer.append(s);
    } catch (IOException e) {
      failed = true;
      logger.atSevere().withCause(e).log(
          "Could not write to file: %s. Redirecting will be disabled.", filePath);
    }
  }

  private void flush() {
    if (failed) {
      return;
    }
    Writer writer = getWriter();
    if (writer != null) {
      try {
        writer.flush();
      } catch (IOException e) {
        failed = true;
        logger.atSevere().withCause(e).log(
            "Could not write to file: %s. Redirecting will be disabled.", filePath);
      }
    }
  }

  private synchronized Writer getWriter() {
    if (writer == null) {
      writer = initWriter();
    }
    return writer;
  }

  @Nullable
  private BufferedWriter initWriter() {
    try {
      StandardOpenOption openOption =
          Files.exists(filePath)
              ? StandardOpenOption.TRUNCATE_EXISTING
              : StandardOpenOption.CREATE_NEW;
      return Files.newBufferedWriter(filePath, StandardCharsets.UTF_8, openOption);
    } catch (IOException e) {
      failed = true;
      logger.atSevere().withCause(e).log(
          "Could not open file: %s. Redirecting will be disabled.", filePath);
    }
    return null;
  }

  @Override
  public void close() {
    super.close();
    synchronized (loggingExecutor) {
      loggingExecutor.shutdown();
      if (flushingExecutor != null) {
        flushingExecutor.shutdown();
      }
    }
    try {
      loggingExecutor.awaitTermination(10, TimeUnit.SECONDS);
      if (flushingExecutor != null) {
        flushingExecutor.awaitTermination(10, TimeUnit.SECONDS);
      }
    } catch (InterruptedException e) {
      logger.atSevere().withCause(e).log("Exception while shutting down.");
    }
    try {
      if (writer == null) {
        return;
      }
      writer.close();
      logger.atInfo().log("Closed file %s", filePath);
    } catch (IOException e) {
      failed = true;
      logger.atSevere().withCause(e).log(
          "Could not close file: %s. Redirecting will be disabled.", filePath);
    }
    loggingExecutor.shutdownNow();
    if (flushingExecutor != null) {
      flushingExecutor.shutdownNow();
    }
  }
}

File: src/Copybara.Core/Util/Console/FileConsole.cs
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Globalization;
using System.Text;
using Copybara.Common;
using MessageType = Copybara.Util.Console.Message.MessageType;

namespace Copybara.Util.Console;

/// <summary>
/// A <see cref="Console"/> capable of writing the output into a file and to a delegate console.
///
/// <para>If any of the file operations fail, the console won't try to write to the file anymore and
/// will only call to the delegate console.</para>
///
/// <para>Caller is responsible for closing this console to free resources.</para>
///
/// <para>The console can be configured to flush on a fixed rate intervals.</para>
/// </summary>
public class FileConsole : DelegateConsole
{
    private const string DatePrefixFmt = "MMdd HH:mm:ss.fff";

    protected readonly string FilePath;
    private readonly Timer? _flushingTimer;

    // Serializes writes to the underlying writer, standing in for the Java single-thread executor.
    private readonly object _loggingLock = new();
    private readonly object _writerLock = new();

    private bool _shutdown = false;
    private bool _failed = false;
    private TextWriter? _writer;

    /// <summary>Creates a new <see cref="FileConsole"/>.</summary>
    /// <param name="delegate">A delegate console.</param>
    /// <param name="filePath">
    /// A file path to write to. The parent directories must be created in advance.
    /// </param>
    /// <param name="consoleFlushRate">How often to flush this file console.</param>
    public FileConsole(Console @delegate, string filePath, TimeSpan consoleFlushRate)
        : base(@delegate)
    {
        FilePath = Preconditions.CheckNotNull(filePath);
        if (consoleFlushRate <= TimeSpan.Zero)
        {
            _flushingTimer = null;
        }
        else
        {
            _flushingTimer = new Timer(_ => Flush(), null, consoleFlushRate, consoleFlushRate);
        }
    }

    protected override void HandleMessage(MessageType type, string message)
    {
        TextWriter? writer = GetWriter();
        if (writer == null)
        {
            return;
        }
        // Submitting to a shut-down executor throws, so this has to be synchronized to avoid shutting
        // down in-between from another thread.
        lock (_loggingLock)
        {
            if (_shutdown)
            {
                return;
            }
            string now = DateTime.Now.ToString(DatePrefixFmt, CultureInfo.InvariantCulture);
            DoWrite(writer, $"{now} {type}: {message}\n");
        }
    }

    private void DoWrite(TextWriter writer, string s)
    {
        if (_failed)
        {
            return;
        }
        try
        {
            writer.Write(s);
        }
        catch (IOException)
        {
            _failed = true;
        }
    }

    private void Flush()
    {
        if (_failed)
        {
            return;
        }
        TextWriter? writer = GetWriter();
        if (writer != null)
        {
            try
            {
                lock (_loggingLock)
                {
                    writer.Flush();
                }
            }
            catch (IOException)
            {
                _failed = true;
            }
        }
    }

    private TextWriter? GetWriter()
    {
        lock (_writerLock)
        {
            if (_writer == null)
            {
                _writer = InitWriter();
            }
            return _writer;
        }
    }

    private TextWriter? InitWriter()
    {
        try
        {
            // TRUNCATE_EXISTING if the file exists, otherwise CREATE_NEW: FileMode.Create matches
            // both cases (create or truncate).
            var stream = new FileStream(FilePath, FileMode.Create, FileAccess.Write, FileShare.Read);
            return new StreamWriter(stream, new UTF8Encoding(false));
        }
        catch (IOException)
        {
            _failed = true;
        }
        return null;
    }

    public override void Dispose()
    {
        base.Dispose();
        lock (_loggingLock)
        {
            _shutdown = true;
        }
        _flushingTimer?.Dispose();
        try
        {
            if (_writer == null)
            {
                return;
            }
            _writer.Dispose();
        }
        catch (IOException)
        {
            _failed = true;
        }
    }
}
